firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/partitioning.ts

```diff
@@ -1,368 +1,283 @@
-import { FirestoreBigQueryEventHistoryTrackerConfig } from ".";
-import { FirestoreDocumentChangeEvent } from "..";
-import * as firebase from "firebase-admin";
+// src/partitioning.ts
 
+import { z } from "zod";
+import { FirestoreBigQueryEventHistoryTrackerConfig } from "./types"; // Assumes public types are in ./types
+import { ChangeType, FirestoreDocumentChangeEvent } from "..";
+import * as firebase from "firebase-admin";
 import * as logs from "../logs";
 import * as bigquery from "@google-cloud/bigquery";
 import * as functions from "firebase-functions";
-import { getNewPartitionField } from "./schema";
-import { BigQuery, TableMetadata } from "@google-cloud/bigquery";
-
-import { PartitionFieldType } from "../types";
-
-export class Partitioning {
-  public config: FirestoreBigQueryEventHistoryTrackerConfig;
-  public table: bigquery.Table;
-  public schema: object;
-
-  constructor(
-    config: FirestoreBigQueryEventHistoryTrackerConfig,
-    table?: bigquery.Table,
-    schema?: object
+import { BigQuery, TableField } from "@google-cloud/bigquery";
+
+const BaseSchema = z.object({
+  timePartitioning: z.any(),
+  timePartitioningColumn: z.any(),
+});
+
+const NoPartitioningInternalSchema = BaseSchema.extend({
+  partitioningType: z.literal("NONE"),
+});
+const IngestionTimeInternalSchema = BaseSchema.extend({
+  partitioningType: z.literal("INGESTION"),
+  timePartitioning: z.enum(["HOUR", "DAY", "MONTH", "YEAR"]),
+});
+const TimestampInternalSchema = BaseSchema.extend({
+  partitioningType: z.literal("TIMESTAMP"),
+  timePartitioning: z.enum(["HOUR", "DAY", "MONTH", "YEAR"]),
+  timePartitioningColumn: z.literal("timestamp"),
+});
+const FieldInternalSchema = BaseSchema.extend({
+  partitioningType: z.literal("FIELD"),
+  timePartitioning: z.enum(["HOUR", "DAY", "MONTH", "YEAR"]),
+  timePartitioningColumn: z.string().min(1),
+  timePartitioningFieldType: z.enum(["TIMESTAMP", "DATE", "DATETIME"]),
+  timePartitioningFirestoreField: z.string().min(1),
+});
+
+const DiscriminatedSchema = z.discriminatedUnion("partitioningType", [
+  NoPartitioningInternalSchema,
+  IngestionTimeInternalSchema,
+  TimestampInternalSchema,
+  FieldInternalSchema,
+]);
+
+const InternalPartitioningSchema = z.preprocess((data) => {
```

Comment:
This adds the partitioningType onto the config...

I dont know whether we actually need the rest of the config, perhaps it's better just to extract a smaller config. Need to check what we actually use here from config